### ISSUE TITLE
PERL_REMOVE_OP_NULL

### ISSUE THREAD

**Issue by rurban** [perl11/cperl#216]
remove NULL ops in rpeep, where the ops are already linearized.
NULL ops in the head are already pruned, just intermediate not.
These should have happened 20 years ago already.

From previous experiences only custom sort blocks didn't work with removed NULL ops in the compiler where we did that already. But with 5.25 it doesn't seem to be needed anymore, since sort was refactored. All tests pass.

I added a `#define PERL_REMOVE_OP_NULL` for custom XS which deal with hard-structured optrees, and don't bypass NULL.
I also added a temp. global `PL_count_null_ops` to see the effect with -Dk (compile time optimized).

Note that the peephole optimizer does not really visit all ops, it leaves some holes, so some NULL ops still pass through. - rpeep is an architectural mess, which should have been rewritten properly before being exported and API'ified. It started as a glorified NULLifier, but was then expanded to a real peep, but kept several NULL related states which have nothing to do with it. op_opt, oldop, oldoldop, defer_queue[], ...

But for minor scripts there are still a few hundreds NULL ops removed and only a few dozen are kept.  NULL ops at the beginning are removed with `prune_chain_head()` already.

Benchmarks are below 4%

testing for leaks:
`PERL_DESTRUCT_LEVEL=2 valgrind --leak-check=full ./miniperl -Ilib -Dk t/comp/fold.t`
(without PERL_DESTRUCT_LEVEL=2 global subs are not destructed, leaking some bytes).
```
== optimized NULL ops: 117
==17241== HEAP SUMMARY:
==17241==     in use at exit: 0 bytes in 0 blocks
==17241==   total heap usage: 5,842 allocs, 5,842 frees, 554,512 bytes allocated
==17241==
==17241== All heap blocks were freed -- no leaks are possible
```
which is logical, given that op_free destroys ops structurely, not just linearily. i.e. it is descending into kids.

**Comment by rurban** [comment]
Merged af12c68b42d126cf6eaade09ccd4c9e8d6e4c08f

**Comment by atoomic** [comment]
Custom blocks for sort is pretty common, so my understanding is that we need to wait for 5.26 to do this. 

Do you have a list of XS modules relying on hard structured trees ? Maybe we can patch them to support the NULL removal ?

Is 4% saving for memory or speed or both ?

**Comment by rurban** [comment]
No, this was fixed with 5.22 upstream
